fdbserver/DataDistribution.actor.cpp

```diff
@@ -2353,8 +2385,85 @@ ACTOR Future<Void> teamRemover(DDTeamCollection* self) {
 	}
 }
 
+// Remove the server team whose members have the most number of process teams
+// until the total number of server teams is no larger than the desired number
+ACTOR Future<Void> serverTeamRemover(DDTeamCollection* self) {
+	state int numServerTeamRemoved = 0;
+	loop {
+		// In case the serverTeamRemover cause problems in production, we can disable it
+		if (SERVER_KNOBS->TR_FLAG_DISABLE_SERVER_TEAM_REMOVER) {
+			return Void(); // Directly return Void()
+		}
+
+		wait(waitUntilHealthy(self));
+
+		// To avoid removing machine teams too fast, which is unlikely happen though
+		wait( delay(SERVER_KNOBS->TR_REMOVE_SERVER_TEAM_DELAY) );
+
+		// Wait for the badTeamRemover() to avoid the potential race between adding the bad team (add the team tracker)
+		// and remove bad team (cancel the team tracker).
+		wait(self->badTeamRemover);
+
+		// Q: We may need to count the number of servers instead of only healthy servers, since healthyness can change quickly?
+		state int healthyServerCount = self->teams.size(); //self->calculateHealthyServerCount();
+		// Check if all servers are healthy, if not, we wait for 1 second and loop back.
+		// Eventually, all servers will become healthy.
+		if (healthyServerCount != self->server_info.size()) {
+			continue;
+		}
+
+		// From this point, all server teams should be healthy, because we wait above
+		// until processingUnhealthy is done, and all machines are healthy
+
+		// In most cases, all machine teams should be healthy teams at this point.
+		int desiredServerTeams = SERVER_KNOBS->DESIRED_TEAMS_PER_SERVER * healthyServerCount;
+		int totalSTCount = self->teams.size();
+
+		if (totalSTCount > desiredServerTeams) {
+			// Pick the server team whose members are on the most number of server teams, and mark it undesired
+			state std::pair<Reference<TCTeamInfo>, int> foundSTInfo = self->getServerTeamWithMostProcessTeams(true);
+			if (!foundSTInfo.first.isValid()) {
+				foundSTInfo = self->getServerTeamWithMostProcessTeams(false);
+			}
+			state Reference<TCTeamInfo> st = foundSTInfo.first;
+			state int maxNumProcessTeams = foundSTInfo.second;
```

Comment:
These two don't need to be state vars I think? Don't see any wait in between declaration and use.

Reply:
yeah, you are right. I removed them.